File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void bfs(int startRow, int startCol, vector<vector<int>> &vis, vector<vector<char>>& grid) {
        vis[startRow][startCol] = 1;

        queue<pair<int, int>> q;
        q.push({startRow, startCol});

        int n = grid.size();
        int m = grid[0].size();

        int delRow[] = {-1, 0, 1, 0};
        int delCol[] = {0, 1, 0, -1};

        while(!q.empty()) {
            int row = q.front().first;
            int col = q.front().second;
            q.pop(); 

            for(int i = 0; i < 4; i++) {
                int nrow = row + delRow[i]; 
                int ncol = col + delCol[i]; 

                if (
                    nrow >= 0 && nrow < n &&
                    ncol >= 0 && ncol < m &&
                    grid[nrow][ncol] == '1' &&
                    !vis[nrow][ncol]
                ) {
                    vis[nrow][ncol] = 1;
                    q.push({nrow, ncol});
                }
            } 
        }  
    }

    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) return 0;

        int n = grid.size();
        int m = grid[0].size();
        int count = 0;

        vector<vector<int>> vis(n, vector<int>(m, 0));

        for(int row = 0; row < n; row++)  {
            for(int col = 0; col < m; col++) { 
                if(!vis[row][col] && grid[row][col] == '1') {
                    count++;
                    bfs(row, col, vis, grid);
                }
            } 
        }
        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public void bfs(int startRow, int startCol, int[][] vis, char[][] grid) {
        vis[startRow][startCol] = 1;

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{startRow, startCol});

        int n = grid.length;
        int m = grid[0].length;

        int[] delRow = {-1, 0, 1, 0};
        int[] delCol = {0, 1, 0, -1};

        while (!q.isEmpty()) {
            int[] cell = q.poll();
            int row = cell[0];
            int col = cell[1];

            for (int i = 0; i < 4; i++) {
                int nrow = row + delRow[i];
                int ncol = col + delCol[i];

                if (
                    nrow >= 0 && nrow < n &&
                    ncol >= 0 && ncol < m &&
                    grid[nrow][ncol] == '1' &&
                    vis[nrow][ncol] == 0
                ) {
                    vis[nrow][ncol] = 1;
                    q.add(new int[]{nrow, ncol});
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) return 0;

        int n = grid.length;
        int m = grid[0].length;
        int count = 0;

        int[][] vis = new int[n][m];

        for (int row = 0; row < n; row++) {
            for (int col = 0; col < m; col++) {
                if (vis[row][col] == 0 && grid[row][col] == '1') {
                    count++;
                    bfs(row, col, vis, grid);
                }
            }
        }
        return count;
    }
}